compiler/rustc_builtin_macros/src/deriving/generic/mod.rs

```diff
@@ -607,28 +609,13 @@ impl<'a> TraitDef<'a> {
 
         // and similarly for where clauses
         where_clause.predicates.extend(generics.where_clause.predicates.iter().map(|clause| {
-            match *clause {
-                ast::WherePredicate::BoundPredicate(ref wb) => {
-                    ast::WherePredicate::BoundPredicate(ast::WhereBoundPredicate {
-                        span: self.span,
-                        bound_generic_params: wb.bound_generic_params.clone(),
-                        bounded_ty: wb.bounded_ty.clone(),
-                        bounds: wb.bounds.to_vec(),
-                    })
-                }
-                ast::WherePredicate::RegionPredicate(ref rb) => {
-                    ast::WherePredicate::RegionPredicate(ast::WhereRegionPredicate {
-                        span: self.span,
-                        lifetime: rb.lifetime,
-                        bounds: rb.bounds.to_vec(),
-                    })
-                }
-                ast::WherePredicate::EqPredicate(ref we) => {
+            match clause {
+                ast::WherePredicate::BoundPredicate(_)
+                | ast::WherePredicate::RegionPredicate(_) => clause.clone(),
+                ast::WherePredicate::EqPredicate(we) => {
                     ast::WherePredicate::EqPredicate(ast::WhereEqPredicate {
                         id: ast::DUMMY_NODE_ID,
```

Comment:
I'm pretty sure that the `clause`'s ID is also dummy, so the predicates can be extended with just `generics.where_clause.predicates.iter().cloned()`.

Reply:
There's one more thing though - the person who previously wrote this clearly didn't know about hygiene.

If you generate code from a derive macro, you cannot reuse the input spans directly.
You can keep the locations, but you have to add one more macro layer to the hygiene parts of the spans.
This is done through the construction `inherited_span.with_ctxt(self.span)` or `self.span.with_span_pos(inherited_span)`, see e.g. `fn typaram`.
The existing code uses the correct hygiene (but with suboptimal position) by using just `self.span`, but only for the top level, all the nested spans are still cloned unmodified.
This PR makes things worse by cloning everything in where clauses unmodified, including the top level.
What is needed here is a small mutable visitor deeply updating all the spans in the original where clauses.

(Other cloned pieces of AST like generic parameters also need to be double-checked for this issue.)

Reply:
I'll annotate all the spans I find with a new context (as appropriate).

Reply:
Changed. I left the cases where `trait_.span` is used under the assumption that those are already correctly pointing at the `derive` name `Symbol` with the appropriate context.

Reply:
The spans are still not updated deeply in the cloned where clauses, for example `wb.bounded_ty`, `we.(l,r)hs_ty` will keep the original spans.

>What is needed here is a small mutable visitor deeply updating all the spans in the original where clauses.

Reply:
Got it. Are there other places where we should use the `MutVisitor` on? It almost feels like once I add this it would be better to yank out most of the span `with_ctxt` markings and replace with it a top level `visit_item` or similar call. Seems less error prone.

Reply:
Usually built-in derives generate entirely new code, not clone existing AST (except for tiny things like identifiers that don't need recursive visiting), so the `where` clause cloning case seems unique enough.

I'm not sure how to do all of this less error prone in centralized way, probably needs some experimentation.
Maybe generating an item first and then visiting it and adjusting all spans is the way to go, it only can cause issues if some parts of the output *don't* need to be adjusted, but with built-in derives it seems unlikely.

Reply:
Could you take a look at the following draft https://github.com/rust-lang/rust/pull/91441?

I'm clearly doing something wrong, as I'm getting failures in the following tests:

```
failures:
    [ui] ui/derives/derive-hygiene.rs
    [ui] ui/hygiene/nested_macro_privacy.rs
    [ui] ui/proc-macro/issue-42708.rs
    [ui] ui/repr/repr-no-niche.rs
    [ui] ui/symbol-names/const-generics-structural-demangling.rs
```

related to proper hygiene when multiple levels of nesting occur. I've tried multiple things but all failed one way or another, so I wanted to check in with you as you might be able to spot some obvious thing I'm missing. I've also pushed the last WIP commit to this PR just in case you wanted to see what else I was doing, but the new draft has *only* the visitor change.